src/main/java/com/github/zly2006/reden/rvc/tracking/TrackedStructure.kt

```diff
@@ -25,16 +34,197 @@ class TrackedStructure (
     override lateinit var world: World
     override val origin: BlockPos.Mutable = BlockPos.ORIGIN.mutableCopy()
     override fun createPlacement(world: World, origin: BlockPos) = this
+    internal var cachedPositions = mutableMapOf<BlockPos, TrackPoint>()
+    internal var cachedIgnoredPositions = mutableMapOf<BlockPos, TrackPoint>()
     val trackPoints = mutableListOf<TrackPoint>()
     val blockEvents = mutableListOf<BlockEvent>() // order sensitive
     val blockScheduledTicks = mutableListOf<NbtCompound>() // order sensitive
     val fluidScheduledTicks = mutableListOf<NbtCompound>() // order sensitive
 
-    class TrackPoint(
+    fun debugRender() {
+        cachedPositions.forEach {
+            if (!world.isAir(it.key))
+                BlockBorder[it.key] = 1
+        }
+        cachedIgnoredPositions.forEach {
+            if (!world.isAir(it.key))
+                BlockBorder[it.key] = 2
+        }
+    }
+
+    val blockBox: BlockBox
+        get() {
+            val minX = cachedPositions.keys.minOf { it.x }
+            val minY = cachedPositions.keys.minOf { it.y }
+            val minZ = cachedPositions.keys.minOf { it.z }
+            val maxX = cachedPositions.keys.maxOf { it.x }
+            val maxY = cachedPositions.keys.maxOf { it.y }
+            val maxZ = cachedPositions.keys.maxOf { it.z }
+            return BlockBox(minX, minY, minZ, maxX, maxY, maxZ)
+        }
+
+    fun splitCuboids(
+        untrackedIsIgnored: Boolean = false,
+    ): List<BlockBox> {
+        class IncludeEntry(
+            var cuboid: BlockBox?,
+            val points: SortedSet<BlockPos> = sortedSetOf()
+        ) {
+            constructor(points: Collection<BlockPos>) : this(
+                null,
+                points.toSortedSet()
+            ) {
+                shrinkCuboid()
+            }
+            fun shrinkCuboid() {
+                if (points.isEmpty()) return
+                val minX = points.minOf { it.x }
+                val minY = points.minOf { it.y }
+                val minZ = points.minOf { it.z }
+                val maxX = points.maxOf { it.x }
+                val maxY = points.maxOf { it.y }
+                val maxZ = points.maxOf { it.z }
+                cuboid = BlockBox(minX, minY, minZ, maxX, maxY, maxZ)
+            }
+        }
+        val result: MutableList<IncludeEntry>
+
+        if (untrackedIsIgnored)
+            result = (cachedPositions.keys).map { IncludeEntry(listOf(it)) }.toMutableList()
+        else {
+            result = mutableListOf(IncludeEntry(cachedPositions.keys))
+            cachedIgnoredPositions.forEach { ignored ->
```

Comment:
Suggest changing `ignored` to `position`. The former can cause misunderstanding.